src/main/java/de/maxhenkel/reap/TreeEvents.java

```diff
@@ -59,97 +64,120 @@ public static boolean canHarvest(BlockPos pos, Player player, Level world, ItemS
-        public boolean contains(Object o) {
-            return stream().anyMatch(pos1 -> pos1.equals(o));
+        final boolean[] leavesFound = new boolean[1];
+        final LinkedList<BlockPos> result =
+                recursiveSearch(level, start, (pos, bs, isRightBlock) -> {
+                    if (isLeaves(bs)) {
+                        leavesFound[0] = true;
+                    }
+                    return true;
+                });
+        return leavesFound[0] ? result : new LinkedList<>();
+    }
+
+    // for internal use
+    private interface BlockAction {
+        boolean onBlock(BlockPos pos, BlockState state, boolean isRightBlock);
+    }
+
+    // Recursively scan 3x3x3 cubes while keeping track of already scanned blocks to avoid cycles.
+    private static LinkedList<BlockPos> recursiveSearch(final LevelReader world, final BlockPos start, @Nullable final BlockAction action) {
+        final Block wantedBlock = world.getBlockState(start).getBlock();
+        boolean abort = false;
+        final LinkedList<BlockPos> result = new LinkedList<>();
+        final Set<BlockPos> visited = new HashSet<>();
+        final LinkedList<BlockPos> queue = new LinkedList<>();
+        int maxHarvestingCount = Main.SERVER_CONFIG.treeHarvestMaxCount.get();
+        queue.push(start);
+
+        while (!queue.isEmpty()) {
+            final BlockPos center = queue.pop();
+            final int x0 = center.getX();
+            final int y0 = center.getY();
+            final int z0 = center.getZ();
+            for (int z = z0 - 1; z <= z0 + 1 && !abort; ++z) {
+                for (int y = y0 - 1; y <= y0 + 1 && !abort; ++y) {
+                    for (int x = x0 - 1; x <= x0 + 1 && !abort; ++x) {
+                        final BlockPos pos = new BlockPos(x, y, z);
+                        final BlockState bs = world.getBlockState(pos);
+                        if ((bs.isAir() || !visited.add(pos))) {
+                            continue;
+                        }
+                        final boolean isRightBlock = bs.is(wantedBlock);
+                        if (isRightBlock) {
+                            result.add(pos);
+                            if (queue.size() > maxHarvestingCount) {
+                                abort = true;
+                                break;
+                            }
+                            queue.push(pos);
+                        }
+                        if (action != null) {
+                            abort = !action.onBlock(pos, bs, isRightBlock);
+                        }
+                    }
+                }
+            }
         }
+        return !abort ? result : new LinkedList<>();
     }
 
+    // leaves detector
+    // Naturally generated leaves don't have BlockStateProperties.PERSISTENT property, meaning it's a real tree.
+    // If there is BlockStateProperties.PERSISTENT property, it means the leave block is placed by player, meaning it's not a natural tree.
```

Comment:
Please use proper javadocs syntax and if you are referencing fields, use `@link`